### ISSUE TITLE
Enable ML-KEM in FIPS provider

### ISSUE THREAD

**Issue by baentsch** [openssl/openssl#26036]
https://github.com/openssl/openssl/pull/25848 added ML-KEM-768 support to the default provider. This issue is to track requirements and prerequisites to be fulfilled for adding it to the FIPS provider, too. Tagging @paulidale @slontis for their experience on this.

**Comment by paulidale** [comment]
Adding the implementation to the FIPS provider is straightforward:

1. Add the algorithms to `providers/fips/fipsprov.c`.
2. Update the relevant `build.info` files to include the implementations in the common library.
3. Add self tests (this is the troublesome one).

There are more steps required to formally validate the implementations and get algorithm certificates etc but they are mostly handled by the FIPS lab.

**Comment by baentsch** [comment]
Thanks @paulidale . Can you list in more detail (or by reference to specs) what step 3 entails? Steps 1 and 2 seem like equivalents to what's already (being) done for default provider, right?

**Comment by paulidale** [comment]
Yes, steps 1 and 2 are the same as for the default provider.

Step 3 involves adding cases to `providers/fips/self_test_kats.c` and `providers/fips/self_test_data.inc`. There aren't any tests for KEMs currently, so something would need to be added.  The reason there are no KEM tests is that the algorithms that currently can do key encapsulation can also do something else and each algorithm only needs to be tested for one mode of operation.

**Comment by t8m** [comment]
Aren't there also pairwise tests required for the key generation? If it is at all possible to have pairwise tests though.

**Comment by paulidale** [comment]
Yes, pairwise tests are likely required.  They slipped my mind & not for the first time :grin:

**Comment by StephenWall** [comment]
Are new algorithms only going into the current major/minor version FIPS module, or will they also appear if the next 3.0.x FIPS module to be certified?

**Comment by t8m** [comment]
> Are new algorithms only going into the current major/minor version FIPS module, or will they also appear if the next 3.0.x FIPS module to be certified?

No new algorithms will be added to 3.0.x FIPS module. It would not be possible to validate them there.

**Comment by t8m** [comment]
> Does that mean a new certified FIP provider from that version as well?

Which would be the next validated version of the FIPS provider and whether it will contain the post quantum algorithms is not yet determined. Currently we do not expect the post quantum algorithms in 3.5 version being FIPS validated.

**Comment by StephenWall** [comment]
Alright.  Thanks again.

**Comment by baentsch** [comment]
Closing as per https://github.com/openssl/project/issues/1054
